cloudwatch-logs-alerts/lambda/src/error-subscriber.ts

```diff
@@ -0,0 +1,115 @@
+const snsTopicEmailArn = process.env.SNS_TOPIC_EMAIL_ARN;
+const snsTopicChatbotArn = process.env.SNS_TOPIC_CHATBOT_ARN;
+
+// to prevent spamming per source, we keep track of the sources we're handling and clear the flag after a timeout, in the form of { sourceLogGroupName: true, ... }
+const isHandlingTimeout = 1000 * 60; // 1 minute
+const handlingSource: Record<string, boolean> = {};
+
+// Logs that are sent to a receiving service through a subscription filter are base64 encoded and compressed with the gzip format.
+// https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/SubscriptionFilters.html#LambdaFunctionExample
+export const handler = async (event: CloudWatchLogsEvent) => {
+  console.log("[Received Event]:", event);
+
+  let uniqueServiceKey = "";
+
+  try {
+    if (!stage || !primaryRegion || !snsTopicEmailArn) {
+      throw new Error("Required environment variables are missing.");
+    }
+
+    const buffer = Buffer.from(event.awslogs.data, "base64");
+    const decompressedData = gunzipSync(buffer).toString("utf-8");
+    const parsed = JSON.parse(decompressedData);
+    const logGroup = parsed?.logGroup;
+    const logStream = parsed?.logStream;
+
+    if (typeof logGroup !== "string" || typeof logStream !== "string") {
+      throw new Error("Could not parse log group or log stream.");
+    }
+
+    if (!handlingSource[logGroup]) {
+      uniqueServiceKey = logGroup;
+      handlingSource[uniqueServiceKey] = true;
+
+      console.log("[Parsed]:", parsed);
+      const message =
+        parsed?.logEvents[0]?.message ?? "Message could not be parsed.";
+      const messageString = JSON.stringify(message, null, 2);
+      console.log("[Message]:", messageString);
+
+      const subject = getSubject(logGroup);
+      const dashboardUrl = getDashboardUrl(subject);
+      const logGroupRegion = resolveEventRegion(parsed?.subscriptionFilters);
+      const logEventsUrl = getLogEventsUrl(logGroupRegion, logGroup, logStream);
+      let emailMessage = `${transformTextToMarkdown(
+        messageString
+      )}\n\nView in AWS console: ${logEventsUrl}}`;
+
+      if (dashboardUrl) {
+        emailMessage += `\n\nView dashboard: ${dashboardUrl}`;
+      }
+
+      // for chatbot / slack integration, custom format needed
+      const chatbotCustomFormat = getChatbotCustomFormat({
+        subject,
+        message: messageString,
+        logEventsUrl,
+        dashboardUrl,
+      });
+
+      // sns promises, include chatbot sns topic if configured
+      const snsPromises = [
+        publishSnsMessage(snsTopicEmailArn, subject, emailMessage),
+        ...(snsTopicChatbotArn
+          ? [
+              publishSnsMessage(
+                snsTopicChatbotArn,
+                subject,
+                JSON.stringify(chatbotCustomFormat)
+              ),
+            ]
+          : []),
+      ];
+
+      // publish to sns topics
+      await Promise.all(snsPromises);
+
+      // clear flag after timeout
```

Comment:
En ole ihan satavarma, mutta tää ei ehkä toimi siten kuin voisi ymmärtää tarkoituksen olevan. Tässä sellainen että toi taitaa toi setTimeout "jäätyä" aina kun ei ole invoke meneillään eli se kyllä taitaa laskea aikaa alaspäin mutta ei käytännössä ole "1 minuutti", muutoin kuin silloin kun invokea tulee jatkuvasti.. eli se taitaa toimia kuten halutaan mutta ei kuten voisi lukeamalla nopeasti ymmärtää niinkö!

Reply:
"The timeout running on the event loop is frozen, and it continues running the next time the function is invoked." 
- https://docs.aws.amazon.com/lambda/latest/dg/nodejs-handler.html

Reply:
Hm, empiirisesti testattuna kyl se näyttäisi siellä kuitenkin ajautuvan ja ajallaan. Ehkäpä toi dokkarin lainaus liittyi vain tuohon vanhaan lambdojen patterniin että käyttelee handlerin kolmatta parametria eli callbackkia.

Reply:
tämä kyllä omien testauksien perusteella kanssa toiminut